packages/components/CONTRIBUTING.md

```diff
@@ -4,12 +4,133 @@ Thank you for taking the time to contribute.
 
 The following is a set of guidelines for contributing to the `@wordpress/components` package to be considered in addition to the general ones described in our [Contributing Policy](/CONTRIBUTING.md).
 
-## Examples
+## Core principles
 
-Each component needs to include an example in its README.md file to demonstrate the usage of the component.
+Contributions to the `@wordpress/components` package should follow a set of core principles and technical requirements.
 
-These examples can be consumed automatically from other projects in order to visualize them in their documentation. To ensure these examples are extractable, compilable and renderable, they should be structured in the following way:
+This set of guidelines should apply especially to newly introduced components. It is, in fact, possible that some of the older component don't respect some of these guidelines for legacy/compat reasons.
 
--   It has to be included in a `jsx` code block.
--   It has to work out-of-the-box. No additional code should be needed to have working the example.
--   It has to define a React component called `My<ComponentName>` which renders the example (i.e.: `MyButton`). Examples for the Higher Order Components should define a `MyComponent<ComponentName>` component (i.e.: `MyComponentWithNotices`).
+### Compatibility
+
+The `@wordpress/components` package includes several components which are relied upon by many developers across different projects. It is, therefore, very important to avoid introducing breaking changes as much as possible.
+
+In these situations, an alternative approach is to "soft-deprecate" the given legacy API. This is achieved by removing traces of the API from the docs, while the code keeps supporing it in the background (with or without a warning)
+
+### Components structure
+
+The contents of the `@wordpress/components` package can be divided into three groups: UI Primitives, UI Blocks, and Utilities.
+
+#### UI Primitives
+
+Lower-level components, which are usually composed together to build the second type of components: UI Blocks. UI Primitives usually don't contain translated text and aim at being highly configurable (at the expense of exposing more props and therefore more complexity).
```

Comment:
@mtias has different terms for this same concept I believe.

Reply:
> at the expense of exposing more props and therefore more complexity

This worries me a bit, we could quickly regret it 🤷‍♂️ I think if we can keep these "primitives" internal, it's better (not expose them in wp.components).

Also, we already have a concept of "primitives": components that replace DOM elements to be mobile compatible `@wordpress/primitives`

Reply:
> @mtias has different terms for this same concept I believe.
> Also, we already have a concept of "primitives"

I'm completely fine with using different terminology (rather than `Primitives` and `UI blocks`). The most important thing to me is that we (if we want to) introduce a distinction between low-level and high-level components, which may help users of this package (including other Gutenberg developers) to choose the right component for a given task.

Does anyone have any suggestions on the naming?

> This worries me a bit, we could quickly regret it 🤷‍♂️

I'm fine with removing the `at the expense of exposing more props and therefore more complexity` text — the main reason why I added it is because it had been previously suggested as a potential difference between "low" and "high" level components.

> I think if we can keep these "primitives" internal, it's better (not expose them in wp.components).

I think that low-level components can still be useful — for example, in the case a user wanted to create their own version of a UI component, but without re-implementing the functionality from scratch and while keeping the WordPress look&feel.

Reply:
This topic can be discussed at a later time in this dedicated issue: #34305

In the meantime, I will go ahead and remove any text referencing any potential components classification in this PR